compiler/src/dotty/tools/dotc/core/Contexts.scala

```diff
@@ -1013,4 +1020,85 @@ object Contexts {
       if (thread == null) thread = Thread.currentThread()
       else assert(thread == Thread.currentThread(), "illegal multithreaded access to ContextBase")
   }
-}
+  end ContextState
+
+  /** Collect information about the run for purposes of additional diagnostics.
+   */
+  class Usages:
+    import rewrites.Rewrites.patch
+    private val selectors   = mutable.Map.empty[ImportInfo, Set[untpd.ImportSelector]].withDefaultValue(Set.empty)
+    private val importInfos = mutable.Map.empty[CompilationUnit, List[(ImportInfo, Symbol)]].withDefaultValue(Nil)
+
+    // register an import
+    def +=(info: ImportInfo)(using Context): Unit =
+      if ctx.settings.WunusedHas.imports && !ctx.owner.is(Enum) && !ctx.compilationUnit.isJava then
+        importInfos(ctx.compilationUnit) ::= ((info, ctx.owner))
+
+    // mark a selector as used
+    def use(info: ImportInfo, selector: untpd.ImportSelector)(using Context): Unit =
+      if ctx.settings.WunusedHas.imports && !info.isRootImport then
+        selectors(info) += selector
+
+    // unused import, owner, which selector
+    def unused(using Context): List[(ImportInfo, Symbol, untpd.ImportSelector)] =
+      var unusages = List.empty[(ImportInfo, Symbol, untpd.ImportSelector)]
+      if ctx.settings.WunusedHas.imports && !ctx.compilationUnit.isJava then
+        //if ctx.settings.Ydebug.value then
+        //  println(importInfos.get(ctx.compilationUnit).map(iss => iss.map((ii, s) => s"${ii.show} ($ii)")).getOrElse(Nil).mkString("Registered ImportInfos\n", "\n", ""))
+        //  println(selectors.toList.flatMap((k,v) => v.toList.map(sel => s"${k.show} -> $sel")).mkString("Used selectors\n", "\n", ""))
+        def checkUsed(info: ImportInfo, owner: Symbol): Unit =
+          val used = selectors(info)
+          var needsPatch = false
+          def cull(toCheck: List[untpd.ImportSelector]): Unit =
+            toCheck match
+              case selector :: rest =>
+                cull(rest) // reverse
+                if !selector.isMask && !used(selector) then
+                  unusages ::= ((info, owner, selector))
+                  needsPatch = true
+              case _ =>
+          cull(info.selectors)
```

Comment:
`cull(info.selectors)` seems to be mis-indented :eyes:

Reply:
Maybe github will add a feature to show vertical lines for indentations.

Reply:
that would be nice, in the meantime I'll buy new glasses I think :smile: